File: 01_ML/01_supervised/01_regression/02_nonlinear_reg/python/02b_cubic.py
"""ML: Non-Linear Regression - Polynomial / Cubic.

Neighborhood Quality Index (0-100) vs price: rough zones are flat-low,
working-class areas rise slowly, middle/upper-middle suburbs climb steeply,
luxury enclaves surge — an S-curve a cubic captures.

    price = a * nqi^3 + b * nqi^2 + c * nqi + d

Run:
    python 02b_cubic.py
"""

import matplotlib

matplotlib.use("Agg")

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures

PLOT_PATH = "02b_cubic_plot.png"


def main() -> None:
    data = {
        "nqi":   [10,  20,  30,  40,  50,  60,  70,  80,  85,  90,  95,  100],
        "price": [180, 195, 215, 250, 310, 410, 560, 760, 900, 1080, 1320, 1600],
    }
    test = {"nqi": [25, 55, 75, 92]}

    df = pd.DataFrame(data)
    print("training data (price in $k):")
    print(df.to_string(index=False))

    poly = PolynomialFeatures(degree=3, include_bias=False)
    X = poly.fit_transform(df[["nqi"]].values)
    reg = LinearRegression().fit(X, df["price"])

    # PolynomialFeatures returns columns in order [x, x^2, x^3]
    c1, c2, c3 = reg.coef_
    d = reg.intercept_
    print(f"\nfit: price = {c3:.4f} * nqi^3 + {c2:.4f} * nqi^2 + {c1:.4f} * nqi + {d:.4f}")

    tst = pd.DataFrame(test)
    tst["price"] = reg.predict(poly.transform(tst[["nqi"]].values))
    print("\npredictions:")
    print(tst.to_string(index=False))

    nqi_grid = np.linspace(df.nqi.min(), df.nqi.max(), 200)
    curve = reg.predict(poly.transform(nqi_grid.reshape(-1, 1)))

    plt.figure()
    plt.xlabel("Neighborhood Quality Index")
    plt.ylabel("price ($k)")
    plt.scatter(df.nqi, df.price, color="magenta", marker="*", s=80, label="training")
    plt.plot(nqi_grid, curve, color="blue", label="cubic fit")
    plt.scatter(tst.nqi, tst.price, color="red", label="predicted")
    plt.legend()
    plt.savefig(PLOT_PATH, bbox_inches="tight")
    print(f"\nWrote {PLOT_PATH}")


if __name__ == "__main__":
    main()

File: 01_ML/01_supervised/01_regression/02_nonlinear_reg/nodejs/02b_cubic.js
// ML: Non-Linear Regression - Polynomial / Cubic
// (Node.js port of ../jupyter/02b_cubic.ipynb)
//
// Neighborhood Quality Index (0-100) vs price: S-curve — flat in rough zones,
// rising through working/middle-class, surging in luxury enclaves.
//
//   price = a * nqi^3 + b * nqi^2 + c * nqi + d

import MLR from 'ml-regression-multivariate-linear';
import { writePlotlyHtml, linspace } from './plot.js';

const data = {
  nqi:   [10, 20, 30, 40, 50, 60, 70, 80, 85, 90, 95, 100],
  price: [180, 195, 215, 250, 310, 410, 560, 760, 900, 1080, 1320, 1600],
};
const test = { nqi: [25, 55, 75, 92] };

console.table(data.nqi.map((n, i) => ({ nqi: n, price: data.price[i] })));

// Cubic features: [nqi, nqi^2, nqi^3]
const X = data.nqi.map((x) => [x, x * x, x * x * x]);
const Y = data.price.map((p) => [p]);

const reg = new MLR(X, Y);

const c1 = reg.weights[0][0];
const c2 = reg.weights[1][0];
const c3 = reg.weights[2][0];
const d  = reg.weights[3][0];
console.log(
  `fit: price = ${c3.toFixed(4)} * nqi^3 + ${c2.toFixed(4)} * nqi^2 + ${c1.toFixed(4)} * nqi + ${d.toFixed(4)}`,
);

const Xtest = test.nqi.map((x) => [x, x * x, x * x * x]);
const predictions = reg.predict(Xtest).map((row) => row[0]);
console.table(test.nqi.map((x, i) => ({ nqi: x, price: predictions[i] })));

const nqiGrid = linspace(Math.min(...data.nqi), Math.max(...data.nqi), 200);
const curve = nqiGrid.map((x) => reg.predict([[x, x * x, x * x * x]])[0][0]);

writePlotlyHtml(
  new URL('./02b_cubic_plot.html', import.meta.url).pathname,
  'Cubic Regression - NQI vs price',
  [
    { type: 'scatter', mode: 'markers', name: 'training',
      x: data.nqi, y: data.price, marker: { color: 'magenta', symbol: 'star', size: 12 } },
    { type: 'scatter', mode: 'lines', name: 'cubic fit',
      x: nqiGrid, y: curve, line: { color: 'blue' } },
    { type: 'scatter', mode: 'markers', name: 'predicted',
      x: test.nqi, y: predictions, marker: { color: 'red', size: 10 } },
  ],
  { xaxis: { title: 'Neighborhood Quality Index' }, yaxis: { title: 'price ($k)' } },
);

console.log('\nWrote 02b_cubic_plot.html');
